rpc_jobs/setup_infra.yml

```diff
@@ -0,0 +1,133 @@
+                )
+                common.venvGalaxy(
+                  "install",
+                  "-r ${{env.WORKSPACE}}/rpc-gating/{name_lower}/role_requirements.yml",
+                  "-p ${{env.WORKSPACE}}/roles"
+                )
+              }}
+            }} // stage
+
+            stage("Prepare Environment"){{
+
+              // write the clouds.yaml file
+              env.OS_CLIENT_CONFIG_FILE = common.writeCloudsCfg()
+
+              // Tell ansible where to find roles
+              env.ANSIBLE_ROLES_PATH = "${{env.WORKSPACE}}/roles"
+
+              // Tell ansible where the dynamic inventory is
+              env.ANSIBLE_INVENTORY = "${{env.WORKSPACE}}/rpc-gating/scripts/ansible_v2_3_2_0_1_contrib_inventory_openstack.py"
+
+              // Tell the dynamic inventory script to only use the 'public_cloud' cloud
+              env.OS_CLOUD = "public_cloud"
+
+              // Tell ansible where the ssh private key is
+              env.ANSIBLE_PRIVATE_KEY_FILE = "${{env.JENKINS_SSH_PRIVKEY}}"
+
+            }} // stage
+
+            stage("Prepare Instances"){{
+              dir('rpc-gating/playbooks'){{
+
+                // Run the setup playbook
+                common.venvPlaybook(
+                  playbooks: [
+                    "setup_openstack_instances.yml"
+                  ],
+                  args: [
+                    "-v",
+                    "-e '@${{env.WORKSPACE}}/rpc-gating/{name_lower}/instance_list.yml'"
+                  ]
+                ) //venvPlaybook
+
+                // We need an inventory refresh, so we execute
+                // this using a fresh shell instead of including
+                // the playbooks in the previous ansible-playbook
+                // play list.
+
+                // Run the instance prep plays
+                common.venvPlaybook(
+                  playbooks: [
+                    "drop_ssh_auth_keys.yml",
+                    "slave_security.yml"
+                  ],
+                  args: [
+                    "-v",
+                    "-e target_hosts={name_lower}_server"
+                  ]
+                ) //venvPlaybook
+
+              }} //dir
+            }} // stage
+
+            stage("Setup {name}"){{
+              dir('rpc-gating/{name_lower}'){{
+
+                // Run the setup playbook
+                common.venvPlaybook(
+                  playbooks: [
+                    "setup_{name_lower}.yml"
+                  ],
+                  args: [
+                    "-v"
+                  ]
+                ) //venvPlaybook
+              }} //dir
+            }} // stage
+          }} //withCredentials
+        }} catch (e) {{
+          print(e)
+          throw(e)
+        }} finally {{
+          common.archive_artifacts()
+        }}
+      }} // globalWraps
+
+- project:
+    name: 'Setup-Infra'
+    CRON: "@daily"
+    RPC_GATING_BRANCH: "master"
+    app:
+      - Grafana:
+          name: "Grafana"
```

Comment:
I don't believe `name: "Grafana"` is necessary, in the template you should just need to use `{app}` instead of `{name}`.

Groovy strings have `toLowerCase`. I think it would be preferable to calculate `name_lower` that have a to define a separate key value pair.

I think this could just be written as:

```yaml
app:
  - Grafana:
      credentials: "id_rsa_cloud10_jenkins_file,RE_GRAFANA_ADMIN_PASSWORD,RE_GRAFANA_GRAFYAML_API_KEY"
```

Reply:
Indeed - updated to use {app} instead of {name} which removes the name parameter.  Nice cleanup.

Also, sadly the templating engine doesn't appear to allow functions (unlike helm where you could easily do `{{ mystring | lower }}`. If I'm wrong, please show me the docs.  Yes, we could use toLowerCase in the groovy DSL block on the strings, but I'd have to introduce string concatenation where ${{env.WORKSPACE}} was used - if only to protect against lowering the workspace directory name with mixed case. I see what you are suggesting, but it's too much cleverness which adds complexity to reduce the obvious parameter. In the present form, it's clear and concise.

Reply:
Why not do something like `name_lower = "{app}".toLowerCase()` and then `foo = "bar_${{name_lower}}_baz"` wherever you want to use it.

Reply:
JJB doesn't use jinja, it does its own macro interpolation 😠, so you can't do translation within jjb. However name_lower is only used within the dsl block, so as @git-harry mentioned you can use groovy string manipulation. (I should have thought of that).